### ISSUE TITLE
Type checking inconsistency: `.pyi` stubs applied for imports but not in implementation file

### ISSUE THREAD

**Issue by koxudaxi** [microsoft/pyright#10241]
## Describe the bug
When using stub files (`.pyi`), Pyright doesn't apply type checking correctly within the implementation file itself, but does apply it correctly when the function is imported in a different file.

## Code or Screenshots
Here's a minimal reproduction of the issue:

### `src/module/__init__.pyi`
```python
from typing import Literal, overload, Callable


@overload
def get_function(version: Literal[1]) -> Callable[[int], int]: ...
@overload
def get_function(version: Literal[2]) -> Callable[[int], str]: ...
@overload
def get_function(version: Literal[3]) -> Callable[[int], float]: ...
```

### `src/module/__init__.py`
```python
from typing import Any, Literal


def get_function(version: int) -> Any:
    # some logic to get the versioned function
    return lambda x: x  # replace with actual function logic


version_2_function = get_function(2)
result: int = version_2_function(10)  # This should be an error because get_function(2) returns Callable[[int], str]
```

### `src/main.py`
```python
from module import get_function


version_2_function = get_function(2)
result: int = version_2_function(10)  # This correctly produces an error
```

## Expected Behavior
Pyright should apply the overload type information from the stub file within the implementation file (`__init__.py`) and produce a type error when assigning a string result to a variable annotated as `int`.

## Actual Behavior
Pyright does not report a type error for the type mismatch within the implementation file, but it does correctly report the error when the function is imported and used in another file.

When running on `src/main.py`:
```
src/main.py:6:15 - error: Type "str" is not assignable to declared type "int"
    "str" is not assignable to "int" (reportAssignmentType)
```

But no errors are reported when running Pyright on the implementation file that contains the exact same code pattern.

## VS Code extension or command-line
I'm running Pyright from the command line, version 1.1.398:
```
$ uvx pyright==1.1.398 src
```

It seems like Pyright does not correctly apply stub type definitions to the same module's implementation file, even though these type definitions should be applied consistently across all files.

**Comment by erictraut** [comment]
Pyright is working correctly here and in conformance with the Python typing spec. Type stubs are intended to be a stand-in for the implementation. They are not applied to the implementation itself. If you want your implementation to be type checked, you need to put the type declarations in the implementation file.

For more details about type stubs and how they are meant to be used, refer to the [Python typing spec](https://typing.python.org/en/latest/spec/distributing.html).

**Comment by koxudaxi** [comment]
Thanks for the clear explanation! I understand now that stub files are meant as stand-ins, not applied to implementation files themselves.
